frontend/app/src/features/room/components/code/CodeEditor.tsx

```diff
@@ -0,0 +1,66 @@
+import { useEffect, useState } from 'react'
+import CodeMirror, { basicSetup } from '@uiw/react-codemirror'
+import { Socket } from "socket.io-client"
+import { LanguageSupport, indentUnit } from '@codemirror/language'
+import { Box, Text } from '@chakra-ui/react'
+import { getDocument, peerExtension } from '@/lib/collab'
+import { Spinner } from '@chakra-ui/react'
+
+interface CodeEditorProps {
+    socket: Socket;
+    className?: string;
+    roomId: string;
+    language: LanguageSupport;
+}
+
+export const CodeEditor = ({ socket, roomId, language }: CodeEditorProps) => {
+
+    const [doc, setDoc] = useState<string | null>(null)
+    const [version, setVersion] = useState<number | null>(null)
+
+    useEffect(() => {
+        const fetchData = async () => {
+            try {
+                const { version, doc } = await getDocument(socket, roomId)
+                setVersion(version)
+                setDoc(doc.toString())
+            } catch (error) {
+                console.error('Error fetching document', error)
+            }
+        }
+
+        fetchData()
+
+        return () => {
+            socket.off('connect')
+            socket.off('disconnect')
+            socket.off('pullUpdateResponse')
+            socket.off('pushUpdateResponse')
+            socket.off('getDocumentResponse')
+        }
+
+    }, [socket, roomId])
+
+
+
+    if (version != null && doc != null) {
+        return (<Box>
+            <Text>{`You are in room ${roomId}`}</Text>
+            <CodeMirror
+                height="100%"
+                basicSetup={false}
+                id="codeEditor"
+                theme="dark"
+                extensions={[
+                    indentUnit.of("\t"),
+                    basicSetup(),
+                    language,
+                    peerExtension(socket, version)
+                ]}
+                value={doc}
+            />
+        </Box>)
+    } else {
+        return <Spinner />
```

Comment:
can help to create a 2 new states
`isLoading` and `error`

so when it starts fetching, it will make `isLoading` true and when `isLoading` is true, the spinner should be rendered

however, if the fetch fails, the `error` will be set to the error message, then the error message should be rendered

if the fetching has finished (i.e. `isLoading` is false) and there is no error (i.e. `error` is undefined/null), then render the code mirror stuff accordingly

Reply:
Added in [d6180e3](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g03/pull/18/commits/d6180e3b696144317f2e611894f7d3418f80b2c8)